File: src/ATBootstrap.jl
module ATBootstrap

using CSV, DataFrames, DataFramesMeta, CategoricalArrays
using Random
using GeoStats, Geodesy, ConcaveHull
using Statistics, StatsBase
using Distributions
using LinearAlgebra
using Distances
using NearestNeighbors
using StaticArrays
using ProgressMeter
using RCall
using StatsPlots, StatsPlots.PlotMeasures
using ColorSchemes

export preprocess_survey_data,
    AbstractSurveyDomain,
    TransectRibbons,
    SurveyHull,
    get_survey_grid,
    read_survey_files,
    in_intervals,
    ATSurveyData,
    ScalingClassProblem,
    ATBootstrapProblem,
    resample_df,
    nonneg_lusim,
    nonneg_lusim!,
    nonneg_lumult,
    nonneg_lumult!,
    zdists,
    simulate_nasc,
    solution_domain,
    simulate_class,
    simulate,
    stepwise_error,
    BootSpecs,
    error_labels,
    plot_class_variograms,
    plot_simulated_nasc,
    plot_geosim_stats,
    plot_boot_results,
    plot_error_sources,
    plot_error_source_by_age,
    summarize_bootstrap,
    summarize_stepwise_bootstrap,
    merge_results

include("types.jl")
include("preprocess_survey_data.jl")
include("spatial.jl")
include("mace_ts.jl")
include("mace_age_length.jl")
include("mace_selectivity.jl")
include("mace_length_weight.jl")
include("calibration.jl")
include("scaling.jl")
include("display.jl")

"""
    Extract the bare x and y values from a `Point` as an `SVector`.
"""
svector_coords(pt::Point) = SVector(pt.coords.x.val, pt.coords.y.val)

function simulate_class_iteration(scp::ScalingClassProblem, surveydata::ATSurveyData,
        bs=BootSpecs(), i=1, scaling_sub = @subset(surveydata.scaling, :class .== scp.class),
        surveygrid_coords = svector_coords.(surveydata.grid),
        z0 = mean.(scp.zdists))

    selectivity_function = make_selectivity_function(bs.selectivity)
    scaling_boot = resample_scaling(scaling_sub, bs.resample_scaling)
    apply_selectivity!(scaling_boot, selectivity_function)

    predict_ts = make_ts_function()
    predict_age = make_age_length_function(surveydata.age_length, scp.age_max,
        bs.age_length)
    
    scaling_boot = DataFramesMeta.@transform!(scaling_boot,
        :sigma_bs = to_linear.(predict_ts.(:ts_relationship, :ts_length, bs.predict_ts)),
        :age = predict_age.(:primary_length))
    
    geotrawls = @chain scaling_boot begin
        @by(:event_id, :class = first(:class))
        innerjoin(surveydata.trawl_locations, on=:event_id)
        @select(:x, :y)
        georef((:x, :y))
    end

    all_ages = make_all_ages(scaling_boot, scp.age_max)
    age_weights = pollock_weights_at_age(scaling_boot, surveydata.length_weight,
        all_ages, bs.weights_at_age)
    
    ii = trawl_assignments(surveygrid_coords, 
                svector_coords.(domain(geotrawls)), bs.trawl_assignments)

    nasc = bs.simulate_nasc ? simulate_nasc(scp) : nonneg_lumult(scp.params, z0)
    cal_error_sim = simulate_cal_error(scp.cal_error,  bs.calibration)
    nasc_df = DataFrame(
        nasc = nasc * cal_error_sim,
        event_id = surveydata.trawl_locations.event_id[ii]
    )

    # Special-case processing for bottom-trawl stratum
    if scp.class == "BT"
        # weight scaling data with Nate's nearbottom coefficients
        nearbottom_dict = make_nearbottom_dict(bs.nearbottom_coefs)
        apply_nearbottom_coefficient!(scaling_boot, nearbottom_dict)
        # If species is not pollock, make TS deterministic (variability in backscatter
        # allocation has already been accounted for with nearbottom coefficients)
        scaling_boot = DataFramesMeta.@transform(scaling_boot,
            @byrow :sigma_bs = :species_code == 21740 ? 
                :sigma_bs : to_linear(predict_ts(:ts_relationship, :ts_length, false))
        )
        # remove nearbottom intercept from nasc (from Nate's paper)
        # nasc_df.nasc .-= nearbottom_intercept
        nasc_df.nasc .= max.(nasc_df.nasc, 0) # make sure we don't end up with negative backscatter
    end

    trawl_means_cat = get_trawl_category_means(scaling_boot, scp.aged_species)
    category_nasc = unstack(trawl_means_cat, :event_id, :category, :p_nasc, fill=0)
    category_sigma = @select(trawl_means_cat, :event_id, :category, :sigma_bs)

    df = @chain nasc_df begin
        leftjoin(category_nasc, on=:event_id)
        DataFrames.stack(Not([:event_id, :nasc]),
            variable_name=:category, value_name=:p_nasc)
        leftjoin(category_sigma, on=[:event_id, :category])
        dropmissing()
        DataFramesMeta.@transform(:n = :nasc .* :p_nasc ./ (4π .* :sigma_bs))
        @by(:category,
            :n = sum(skipmissing(:n)) * surveydata.dA)
        DataFramesMeta.@transform(
            :species_code = parse.(Int, first.(split.(:category, "@"))),
            :age = parse.(Int, last.(split.(:category, "@")))
        )
        leftjoin(age_weights, on=[:species_code, :age])
        DataFramesMeta.@transform(
            :biomass = :n .* :weight, 
            :i = i
        )
        @select(:i, :species_code, :age, :category, :n, :biomass)
        @orderby(:i, :species_code, :age, :n, :biomass)
    end
    return df
end


"""
    simulate_class(scp, surveydata; nreplicates=500, bs=BootSpecs())

Do a bootstrap analysis for one acoustic class/scaling stratum using the problem defined
in `scp` and the data in `surveydata`. The number of bootstrap replicates can optionally
be set in `nreplicates`, and the `BootSpecs` object `bs` can be used to specify if any of
the error sources should be omitted (by default all are included).
"""
function simulate_class(scp::ScalingClassProblem, surveydata::ATSurveyData; nreplicates=500, 
        bs=BootSpecs())

    scaling_sub = @subset(surveydata.scaling, :class .== scp.class)
    surveygrid_coords = svector_coords.(surveydata.grid)
    z0 = mean.(scp.zdists)

    println("Bootstrapping $(scp.class)...")
    results = @showprogress map(1:nreplicates) do i
        simulate_class_iteration(scp, surveydata, bs, i, scaling_sub, surveygrid_coords, z0)
    end

    return vcat(results...)
end

"""
    simulate(atbp, surveydata[; nreplicates=500, bs=BootSpecs(),
        report_species=[21740], report_ages=1:first(atbp.class_problems).age_max)

Run a bootstrap uncertainty estimation for each scaling class, based on the
`ATBootstrapProblem` definition in `atbp` and the data in `surveydata`. The
number of bootstrap replicates can optionally be set in `nreplicates`, and the `BootSpecs`
object `bs` can be used to specify if any of the error sources should be omitted (by
default all are included).

By default, abundance and biomass are reported for pollock by age class. Abundances are
calculated for other species (but not biomasses, and not by age, since this code 
doesn't put together the length-age and length-weight tables for them). If you would like
to return the abundance results for other species, you can list their species codes in the
optional `report_species` argument. For *all* species and ages, pass in empty arrays (`[]`)
for `report_species` and `report_ages`.

For pollock, you can limit the ages reported by passing them to the `report_ages` argument.
Note that this just subsets the results post-simulation; to set the maximum age see the 
documentation for `ScalingClassProblem.`
"""
function simulate(atbp::ATBootstrapProblem, surveydata::ATSurveyData; nreplicates=500,
        bs=BootSpecs(), report_species=[21740], report_ages=1:first(atbp.class_problems).age_max)
    
    class_results = map(p -> simulate_class(p, surveydata; nreplicates, bs), atbp.class_problems)
    if length(report_species) == 0
        report_species = unique(surveydata.scaling.species_code)
    end
    if length(report_ages) == 0
        report_ages = [-1; 1:first(atbp.class_problems).age_max]
    end
    results = @chain vcat(class_results...) begin
        @subset(
            in(report_ages).(:age),
            in(report_species).(:species_code)
        )
        @by([:i, :species_code, :age], :n = sum(:n), :biomass = sum(:biomass))
        DataFramesMeta.@transform(:age = replace(:age, -1 => missing))
    end
    return results
end

"""
    stepwise_errors(atbp, surveydata[; nreplicates=500, remove=false])

Run a stepwise quantification of each error source for the survey analysis defined by 
`atbp` and `surveydata`. If `remove=false` (the default), this will add one 
error source at a time, while fixing all others at zero, i.e. treating them as error-free
or deterministic. Use `nreplicates` to set the number of bootstrap replicates.
"""
function stepwise_error(atbp, surveydata; nreplicates=500, remove=false)
    error_sources = string.(fieldnames(BootSpecs))
    colname = remove ? :eliminated_error : :added_error
    results = map(eachindex(error_sources)) do i
        prefix = remove ? "Omitting" : "Adding"
        println("\n$(prefix) $(error_sources[i]) ($(i)/$(length(error_sources)))...")
        errs = fill(remove, length(error_sources))
        errs[i] = !remove
        bs = BootSpecs(errs...)
        res = simulate(atbp, surveydata; bs=bs, nreplicates=nreplicates)
        res[!, colname] .= error_sources[i]
        res
    end
    return vcat(results...)
end

# Ordered, categorical labels for stepwise error results
error_labels = DataFrame(
    added_error = ["calibration", "simulate_nasc", "selectivity", "resample_scaling",
        "drop_trawl", "trawl_assignments", "predict_ts", "age_length", "weights_at_age", "All"],
    error_label = CategoricalArray(
        ["Calibration", "Spatial sampling", "Selectivity", "Resample catches", 
        "Trawl dropping", "Trawl assignment", "TS models", "Age-length", "Length-weight", "All"],
        levels=["Calibration", "Spatial sampling", "Selectivity", "Resample catches", 
        "Trawl dropping", "Trawl assignment", "TS models", "Age-length", "Length-weight", "All"]
    )
)


"""
    `read_survey_files(surveydir)`

Convenience function to read all files in a survey data directory `surveydir`. Assumes
that the directory contains files with the following names:
- acoustics_projected.csv
- trawl_locations_projected.csv
- scaling.csv
- age_length.csv
- length_weight.csv
- surveygrid.csv
These files are produced by running `download_survey.R`, which gets the files from
Macebase and saves them to the survey data directory, followed by `preprocess_survey_data`,
which projects everything geographically and averages it into spatial bins.
"""
function read_survey_files(surveydir)
    acoustics = CSV.read(joinpath(surveydir, "acoustics_projected.csv"), DataFrame)
    trawl_locations = CSV.read(joinpath(surveydir, "trawl_locations_projected.csv"), DataFrame)
    scaling = CSV.read(joinpath(surveydir, "scaling.csv"), DataFrame)
    scaling = DataFramesMeta.@transform(scaling, :sample_correction_scalar = float(:sample_correction_scalar))
    age_length = CSV.read(joinpath(surveydir, "age_length.csv"), DataFrame)
    length_weight = CSV.read(joinpath(surveydir, "length_weight.csv"), DataFrame)
    surveygrid = CSV.read(joinpath(surveydir, "surveygrid.csv"), DataFrame)
    surveygrid = DataFrames.shuffle(surveygrid) # this seems to fix the issue with directional artifacts
    surveygrid =  PointSet(Point(x...) for x in eachrow(surveygrid))
    return (;acoustics, scaling, age_length, length_weight, trawl_locations, surveygrid)
end

function in_intervals(x::Real, intervals)
    for int in intervals
        if int[1] <= x <= int[2]
            return true
        end
    end
    return false
end

function in_intervals(x::AbstractVector, intervals)
    return [in_intervals(xi, intervals) for xi in x]
end


end # module

File: R/ATBootstrap.R
# ATBootstrap R Module
# Converted from Julia ATBootstrap module

# Load required libraries
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(here)
library(sf)
library(sp)
library(gstat)
library(FNN)
library(progress)
library(MASS)
library(forcats)
library(purrr)
library(stringr)

# Global constants
km2nmi <- 1 / 1.852

# Helper functions
svector_coords <- function(point_data) {
  # Extract x,y coordinates from point data
  return(data.frame(x = point_data$x, y = point_data$y))
}

in_intervals <- function(x, intervals) {
  # Check if values are within specified intervals
  result <- rep(FALSE, length(x))
  for (i in seq_along(intervals)) {
    interval <- intervals[[i]]
    result <- result | (x >= interval[1] & x <= interval[2])
  }
  return(result)
}

# BootSpecs class equivalent
BootSpecs <- function(calibration = TRUE, 
                     simulate_nasc = TRUE, 
                     selectivity = TRUE, 
                     resample_scaling = TRUE,
                     nearbottom_coefs = TRUE, 
                     trawl_assignments = TRUE, 
                     predict_ts = TRUE, 
                     age_length = TRUE, 
                     weights_at_age = TRUE) {
  list(
    calibration = calibration,
    simulate_nasc = simulate_nasc,
    selectivity = selectivity,
    resample_scaling = resample_scaling,
    nearbottom_coefs = nearbottom_coefs,
    trawl_assignments = trawl_assignments,
    predict_ts = predict_ts,
    age_length = age_length,
    weights_at_age = weights_at_age
  )
}

# Error labels for stepwise analysis
error_labels <- data.frame(
  added_error = c("calibration", "simulate_nasc", "selectivity", "resample_scaling",
                  "nearbottom_coefs", "trawl_assignments", "predict_ts", "age_length", 
                  "weights_at_age", "All"),
  error_label = factor(
    c("Calibration", "Spatial sampling", "Selectivity", "Resample catches", 
      "Nearbottom coefs", "Trawl assignment", "TS models", "Age-length", 
      "Length-weight", "All"),
    levels = c("Calibration", "Spatial sampling", "Selectivity", "Resample catches", 
               "Nearbottom coefs", "Trawl assignment", "TS models", "Age-length", 
               "Length-weight", "All")
  )
)

# ATSurveyData class
ATSurveyData <- function(acoustics, scaling, age_length, length_weight, 
                        trawl_locations, grid, dA) {
  structure(
    list(
      acoustics = acoustics,
      scaling = scaling,
      age_length = age_length,
      length_weight = length_weight,
      trawl_locations = trawl_locations,
      grid = grid,
      dA = dA
    ),
    class = "ATSurveyData"
  )
}

# ScalingClassProblem class
ScalingClassProblem <- function(class_name, zdists, params, cal_error, 
                               age_max, aged_species) {
  structure(
    list(
      class = class_name,
      zdists = zdists,
      params = params,
      cal_error = cal_error,
      age_max = age_max,
      aged_species = aged_species
    ),
    class = "ScalingClassProblem"
  )
}

# ATBootstrapProblem class
ATBootstrapProblem <- function(class_problems) {
  structure(
    list(class_problems = class_problems),
    class = "ATBootstrapProblem"
  )
}

# Read survey files function
read_survey_files <- function(surveydir) {
  acoustics <- read_csv(file.path(surveydir, "acoustics_projected.csv"))
  trawl_locations <- read_csv(file.path(surveydir, "trawl_locations_projected.csv"))
  scaling <- read_csv(file.path(surveydir, "scaling.csv")) %>%
    mutate(sample_correction_scalar = as.numeric(sample_correction_scalar))
  age_length <- read_csv(file.path(surveydir, "age_length.csv"))
  length_weight <- read_csv(file.path(surveydir, "length_weight.csv"))
  surveygrid <- read_csv(file.path(surveydir, "surveygrid.csv"))
  # Shuffle surveygrid to fix directional artifacts
  surveygrid <- surveygrid[sample(nrow(surveygrid)), ]
  
  return(list(
    acoustics = acoustics,
    scaling = scaling,
    age_length = age_length,
    length_weight = length_weight,
    trawl_locations = trawl_locations,
    surveygrid = surveygrid
  ))
}

# Simulation functions (these would need to be implemented based on the included files)
simulate_nasc <- function(scp) {
  # Placeholder - would need actual implementation from spatial.jl
  warning("simulate_nasc needs to be implemented based on spatial.jl")
  return(rep(1, length(scp$zdists)))
}

nonneg_lumult <- function(params, z0) {
  # Placeholder - would need actual implementation
  warning("nonneg_lumult needs to be implemented")
  return(z0)
}

# Selectivity functions (from mace_selectivity.jl equivalent)
make_selectivity_function <- function(use_selectivity = TRUE) {
  # Placeholder - would need actual implementation
  function(data) {
    if (use_selectivity) {
      warning("Selectivity function needs to be implemented")
    }
    return(data)
  }
}

apply_selectivity <- function(data, selectivity_func) {
  return(selectivity_func(data))
}

# Target strength functions (from mace_ts.jl equivalent)
make_ts_function <- function() {
  # Placeholder - would need actual implementation
  function(ts_relationship, ts_length, predict_ts) {
    warning("TS function needs to be implemented")
    return(rep(1, length(ts_length)))
  }
}

to_linear <- function(db_values) {
  return(10^(db_values/10))
}

# Age-length functions (from mace_age_length.jl equivalent)
make_age_length_function <- function(age_length_data, age_max, use_age_length = TRUE) {
  # Placeholder - would need actual implementation
  function(lengths) {
    warning("Age-length function needs to be implemented")
    return(sample(1:age_max, length(lengths), replace = TRUE))
  }
}

# Length-weight functions (from mace_length_weight.jl equivalent)
make_weight_function <- function(length_weight_data) {
  # Placeholder - would need actual implementation
  function(lengths) {
    warning("Weight function needs to be implemented")
    return(lengths * 0.01) # placeholder conversion
  }
}

# Scaling functions
resample_scaling <- function(scaling_data, resample = TRUE) {
  if (resample) {
    # Bootstrap resample the scaling data
    n <- nrow(scaling_data)
    indices <- sample(1:n, n, replace = TRUE)
    return(scaling_data[indices, ])
  }
  return(scaling_data)
}

# Calibration functions
simulate_cal_error <- function(cal_error, use_calibration = TRUE) {
  if (use_calibration) {
    warning("Calibration error simulation needs to be implemented")
    return(1 + rnorm(1, 0, 0.1)) # placeholder
  }
  return(1)
}

# Near-bottom processing
make_nearbottom_dict <- function(use_coefs = TRUE) {
  # Placeholder - would need actual implementation
  warning("Near-bottom coefficient dictionary needs to be implemented")
  return(list())
}

apply_nearbottom_coefficient <- function(data, coef_dict) {
  warning("Near-bottom coefficient application needs to be implemented")
  return(data)
}

nearbottom_intercept <- 0 # placeholder value

# Trawl assignment functions
trawl_assignments <- function(grid_coords, trawl_coords, assign = TRUE) {
  if (assign) {
    # Use nearest neighbor assignment
    nn_result <- get.knnx(trawl_coords, grid_coords, k = 1)
    return(nn_result$nn.index[, 1])
  }
  # If not assigning, return sequential indices
  return(1:nrow(grid_coords))
}

# Category processing functions
get_trawl_category_means <- function(scaling_data, aged_species, weight_func) {
  warning("Trawl category means calculation needs to be implemented")
  # Placeholder implementation
  return(scaling_data %>%
    mutate(
      category = paste0(species_code, "@", age),
      p_nasc = 1,
      sigma_bs = 1,
      weight = weight_func(primary_length)
    ) %>%
    select(haul_id, category, p_nasc, sigma_bs, weight))
}

# Core simulation function
simulate_class_iteration <- function(scp, surveydata, bs = BootSpecs(), i = 1) {
  # Get scaling data for this class
  scaling_sub <- surveydata$scaling %>% filter(class == scp$class)
  surveygrid_coords <- surveydata$grid
  z0 <- sapply(scp$zdists, mean)
  
  # Apply bootstrap steps
  selectivity_function <- make_selectivity_function(bs$selectivity)
  scaling_boot <- resample_scaling(scaling_sub, bs$resample_scaling)
  scaling_boot <- apply_selectivity(scaling_boot, selectivity_function)
  
  predict_ts <- make_ts_function()
  predict_age <- make_age_length_function(surveydata$age_length, scp$age_max, bs$age_length)
  
  # Transform scaling data
  scaling_boot <- scaling_boot %>%
    mutate(
      sigma_bs = to_linear(predict_ts(ts_relationship, ts_length, bs$predict_ts)),
      age = predict_age(primary_length)
    )
  
  # Create geotrawls equivalent
  geotrawls <- scaling_boot %>%
    group_by(haul_id) %>%
    summarise(class = first(class), .groups = "drop") %>%
    inner_join(surveydata$trawl_locations, by = "haul_id") %>%
    select(haul_id, x, y)
  
  # Trawl assignments
  ii <- trawl_assignments(surveygrid_coords, 
                         geotrawls %>% select(x, y), 
                         bs$trawl_assignments)
  
  # Simulate NASC
  nasc_val <- if (bs$simulate_nasc) {
    simulate_nasc(scp)
  } else {
    nonneg_lumult(scp$params, z0)
  }
  
  cal_error_sim <- simulate_cal_error(scp$cal_error, bs$calibration)
  nasc_df <- data.frame(
    nasc = nasc_val * cal_error_sim,
    haul_id = geotrawls$haul_id[ii]
  )
  
  # Special processing for bottom-trawl stratum
  if (scp$class == "BT") {
    nearbottom_dict <- make_nearbottom_dict(bs$nearbottom_coefs)
    scaling_boot <- apply_nearbottom_coefficient(scaling_boot, nearbottom_dict)
    
    # Adjust TS for non-pollock species
    scaling_boot <- scaling_boot %>%
      mutate(
        sigma_bs = ifelse(species_code == 21740, 
                         sigma_bs, 
                         to_linear(predict_ts(ts_relationship, ts_length, FALSE)))
      )
    
    # Remove nearbottom intercept
    nasc_df$nasc <- nasc_df$nasc - nearbottom_intercept
    nasc_df$nasc <- pmax(nasc_df$nasc, 0)
  }
  
  # Calculate weights and categories
  predict_weight <- make_weight_function(surveydata$length_weight)
  trawl_means_cat <- get_trawl_category_means(scaling_boot, scp$aged_species, predict_weight)
  
  category_nasc <- trawl_means_cat %>%
    select(haul_id, category, p_nasc) %>%
    pivot_wider(names_from = category, values_from = p_nasc, values_fill = 0)
  
  category_sigma <- trawl_means_cat %>% select(haul_id, category, sigma_bs)
  category_weight <- trawl_means_cat %>% select(haul_id, category, weight)
  
  # Final calculations
  df <- nasc_df %>%
    left_join(category_nasc, by = "haul_id") %>%
    pivot_longer(cols = -c(haul_id, nasc), names_to = "category", values_to = "p_nasc") %>%
    left_join(category_sigma, by = c("haul_id", "category")) %>%
    left_join(category_weight, by = c("haul_id", "category")) %>%
    mutate(
      n = nasc * p_nasc / (4 * pi * sigma_bs) * surveydata$dA,
      biomass = n * weight
    ) %>%
    group_by(category) %>%
    summarise(
      n = sum(n, na.rm = TRUE),
      biomass = sum(biomass, na.rm = TRUE),
      .groups = "drop"
    ) %>%
    mutate(
      species_code = as.numeric(str_extract(category, "^[0-9]+")),
      age = as.numeric(str_extract(category, "[0-9]+$")),
      i = i,
      class = scp$class
    ) %>%
    select(class, i, species_code, age, category, n, biomass) %>%
    arrange(i, species_code, age, n, biomass)
  
  return(df)
}

# Main simulation functions
simulate_class <- function(scp, surveydata, nreplicates = 500, bs = BootSpecs()) {
  cat("Bootstrapping", scp$class, "...\n")
  
  pb <- progress_bar$new(
    format = "[:bar] :percent eta: :eta",
    total = nreplicates,
    width = 60
  )
  
  results <- map_dfr(1:nreplicates, function(i) {
    pb$tick()
    simulate_class_iteration(scp, surveydata, bs, i)
  })
  
  results$class <- scp$class
  return(results)
}

simulate <- function(atbp, surveydata, nreplicates = 500, bs = BootSpecs(),
                    report_species = c(21740), 
                    report_ages = 1:atbp$class_problems[[1]]$age_max) {
  
  class_results <- map(atbp$class_problems, ~simulate_class(.x, surveydata, nreplicates, bs))
  
  if (length(report_species) == 0) {
    report_species <- unique(surveydata$scaling$species_code)
  }
  if (length(report_ages) == 0) {
    report_ages <- c(-1, 1:atbp$class_problems[[1]]$age_max)
  }
  
  results <- bind_rows(class_results) %>%
    filter(
      age %in% report_ages,
      species_code %in% report_species
    ) %>%
    group_by(i, species_code, age) %>%
    summarise(
      n = sum(n),
      biomass = sum(biomass),
      .groups = "drop"
    ) %>%
    mutate(age = ifelse(age == -1, NA, age))
  
  return(results)
}

stepwise_error <- function(atbp, surveydata, nreplicates = 500, remove = FALSE) {
  error_sources <- names(BootSpecs())
  colname <- if (remove) "eliminated_error" else "added_error"
  
  results <- map_dfr(seq_along(error_sources), function(i) {
    prefix <- if (remove) "Omitting" else "Adding"
    cat("\n", prefix, error_sources[i], "(", i, "/", length(error_sources), ")...\n")
    
    errs <- rep(remove, length(error_sources))
    errs[i] <- !remove
    names(errs) <- error_sources
    bs <- do.call(BootSpecs, as.list(errs))
    
    res <- simulate(atbp, surveydata, nreplicates = nreplicates, bs = bs)
    res[[colname]] <- error_sources[i]
    res
  })
  
  return(results)
}

# Utility functions that would need to be implemented based on other files
zdists <- function(atbp) {
  warning("zdists function needs to be implemented")
  return(data.frame(survey = character(), class = character(), zdist = numeric()))
}

solution_domain <- function(scp) {
  warning("solution_domain function needs to be implemented")
  return(list(x = numeric(), y = numeric()))
}

# Summary functions
summarize_bootstrap <- function(results, variable) {
  results %>%
    group_by(species_code, age) %>%
    summarise(
      mean = mean(.data[[variable]], na.rm = TRUE),
      median = median(.data[[variable]], na.rm = TRUE),
      sd = sd(.data[[variable]], na.rm = TRUE),
      q025 = quantile(.data[[variable]], 0.025, na.rm = TRUE),
      q975 = quantile(.data[[variable]], 0.975, na.rm = TRUE),
      .groups = "drop"
    )
}

merge_results <- function(results, results_step) {
  # Add "All" category to main results
  results_all <- results %>%
    mutate(added_error = "All")
  
  bind_rows(results_step, results_all)
}

# Placeholder plotting functions
plot_class_variograms <- function(atbp, ...) {
  warning("plot_class_variograms needs to be implemented")
  return(ggplot() + ggtitle("Variograms - Not Implemented"))
}

plot_simulated_nasc <- function(atbp, surveydata, ...) {
  warning("plot_simulated_nasc needs to be implemented")
  return(ggplot() + ggtitle("Simulated NASC - Not Implemented"))
}

plot_geosim_stats <- function(atbp, surveydata, ...) {
  warning("plot_geosim_stats needs to be implemented")
  return(ggplot() + ggtitle("Geosim Stats - Not Implemented"))
}

plot_boot_results <- function(results, ...) {
  warning("plot_boot_results needs to be implemented")
  return(ggplot() + ggtitle("Bootstrap Results - Not Implemented"))
}

plot_error_sources <- function(results, ...) {
  warning("plot_error_sources needs to be implemented")
  return(ggplot() + ggtitle("Error Sources - Not Implemented"))
}

plot_error_source_by_age <- function(results_step, results, variable) {
  warning("plot_error_source_by_age needs to be implemented")
  return(ggplot() + ggtitle("Error Sources by Age - Not Implemented"))
}

# Preprocessing function placeholder
preprocess_survey_data <- function(surveydir, dx = 10.0, ebs = TRUE, log_ranges = NULL) {
  warning("preprocess_survey_data needs to be implemented - see preprocess_survey_data.jl")
  invisible(NULL)
}

cat("ATBootstrap R module loaded successfully.\n")
cat("Note: Many functions are placeholders and need full implementation based on the Julia source files.\n")
cat("Key files to implement:\n")
cat("- types.jl -> Define R classes/structures\n")
cat("- spatial.jl -> Spatial modeling functions\n") 
cat("- mace_*.jl -> MACE-specific functions\n")
cat("- display.jl -> Plotting functions\n")
